Upgraded to ONOS 2.2.2 and using GUI2 (#41)
opennetworkinglab/ngsdn-tutorial: EXERCISE-3.md

**Note:** in theory, there should be no need to kill and restart ONOS. However,
while ONOS supports reloading apps with a modified one, the version of ONOS used

> in this tutorial (2.2.1, the most recent Long Term Support release at the time

Author: is this still the case in 2.2.1?

```
onos> flows any device:leaf1 | grep appId=org.onosproject.core

>     id=100001e5fba59, state=ADDED, bytes=0, packets=0, duration=355, liveType=UNKNOWN, priority=40000, tableId=IngressPipeImpl.acl_table, appId=org.onosproject.core, selector=[ETH_TYPE:arp], treatment=DefaultTrafficTreatment{immediate=[IngressPipeImpl.clone_to_cpu()], deferred=[], transition=None, meter=[], cleared=false, StatTrigger=null, metadata=null}

Reviewer: I'd prefer if you could revert changes to the `flows` command output. The `...` is used to reduce clutter in the instructions

* Press `A` multiple times until you see link stats, in either 
  packets/seconds (pps) or bits/seconds.

> Only the hosts that have been pinged successfully at this point in mininet will

> >be visible.

Reviewer: ```suggestion
> be visible.
```

Link stats are derived by ONOS by periodically obtaining the port counters for
each device. ONOS internally uses gNMI to read port information, including
counters.

> > All of the devices, ports, portstats, links, hosts, flows and groups can be

Reviewer: ```suggestion
> All of the devices, ports (including stats), links, hosts, flows, and groups can be
```

> >displayed through the various views in the web UI.

Reviewer: ```suggestion
> displayed through the various views in the web UI.
```

#### 7a. View Pipeconf data in the web UI

In the ONOS topology view click on one of the Stratum switches (e.g `device:leaf1`)
and the Device Details panel appears. In that panel click on the Pipeconf icon

> (the last one), to open the Pipeconf view for that device.

Reviewer: It would be  good to attach a screenshot of the panel and pipeconf button to click

In the top panel the general Pipeconf data for that device is given.
In the lower panel the tables of the P4 model are listed, with the number of table

> entries and packets matched dynamically updated.

Reviewer: ```suggestion
In the top panel, the general Pipeconf data for that device is given.
In the lower panel, the same tables found in the P4 program are listed, with the number of table
entries and packets matched periodically updated.
```

![onos-gui-pipeconf-leaf1](img/onos-gui-pipeconf-leaf1.png)

> Clicking the table row brings up the details panel, showing bit widths etc.

Reviewer: ```suggestion
Clicking the table row brings up the details panel, showing details of the match field, actions,
action parameter bit widths, etc.
```

  packets/seconds (pps) or bits/seconds.

> Only the hosts that have been pinged successfully at this point in mininet will
> be visible.

Link stats are derived by ONOS by periodically obtaining the port counters for
each device. ONOS internally uses gNMI to read port information, including
counters.

> All of the devices, ports (including stats), links, hosts, flows, and groups can be
> displayed through the various views in the web UI.

#### 7a. View Pipeconf data in the web UI

In the ONOS topology view click on one of the Stratum switches (e.g `device:leaf1`)
and the Device Details panel appears. In that panel click on the Pipeconf icon
(the last one), to open the Pipeconf view for that device.

![device-leaf1-details-panel](img/device-leaf1-details-panel.png)

In the top panel, the general Pipeconf data for that device is given. In the
lower panel, the same tables found in the P4 program are listed, with the number
of table entries and packets matched periodically updated.

![onos-gui-pipeconf-leaf1](img/onos-gui-pipeconf-leaf1.png)

Clicking the table row brings up the details panel, showing details of the match

> filed, actions, action parameter bit widths, etc.

Reviewer: ```suggestion
fields, actions, action parameter bit widths, etc.
```
